core/internal/runsummary/runsummary.go

```diff
@@ -1,250 +1,270 @@
-	name := strings.Join(path, ".")
+	summary.(*metricSummary).Clear()
+}
 
-	types := make([]SummaryType, 0)
+// UpdateSummaries updates metric summaries based on their new values
+// and returns the updates made.
+//
+// The list of updates may be non-empty even on error. An error state
+// may leave the run summary partially updated.
+func (rs *RunSummary) UpdateSummaries(
+	history *runhistory.RunHistory,
+) ([]*service.SummaryItem, error) {
+	var updates []*service.SummaryItem
+	var errs []error
 
-	for pattern, definedMetric := range rs.mh.HackGetDefinedMetrics() {
-		if pattern == name {
-			summary := definedMetric.GetSummary()
-			if summary.GetNone() {
-				return []SummaryType{None}
-			}
-			if summary.GetMax() {
-				types = append(types, Max)
-			}
-			if summary.GetMin() {
-				types = append(types, Min)
-			}
-			if summary.GetMean() {
-				types = append(types, Mean)
-			}
-			if summary.GetLast() {
-				types = append(types, Latest)
+	history.ForEach(
+		func(path pathtree.TreePath, value float64) bool {
+			update, err := rs.updateSummaryFloat(path, value)
+
+			if err != nil {
+				errs = append(errs, err)
 			}
-		}
-	}
-	for pattern, globMetric := range rs.mh.HackGetGlobMetrics() {
-		// match the key against the glob pattern:
-		// note check for no error
-		if match, err := filepath.Match(pattern, name); err == nil && match {
-			summary := globMetric.GetSummary()
-			if summary.GetNone() {
-				return []SummaryType{None}
+			if update != nil {
+				updates = append(updates, update)
 			}
-			if summary.GetMax() {
-				types = append(types, Max)
+
+			return true
+		},
+		func(path pathtree.TreePath, value int64) bool {
+			update, err := rs.updateSummaryInt(path, value)
+
+			if err != nil {
+				errs = append(errs, err)
 			}
-			if summary.GetMin() {
-				types = append(types, Min)
+			if update != nil {
+				updates = append(updates, update)
 			}
-			if summary.GetMean() {
-				types = append(types, Mean)
+
+			return true
+		},
+		func(path pathtree.TreePath, value any) bool {
+			update, err := rs.updateSummaryOther(path, value)
+
+			if err != nil {
+				errs = append(errs, err)
 			}
-			if summary.GetLast() {
-				types = append(types, Latest)
+			if update != nil {
+				updates = append(updates, update)
 			}
-		}
+
+			return true
+		},
+	)
+
+	return updates, errors.Join(errs...)
+}
+
+func (rs *RunSummary) updateSummaryFloat(
+	path pathtree.TreePath,
+	value float64,
+) (*service.SummaryItem, error) {
+	return rs.updateSummary(path, func(ms *metricSummary) {
+		ms.UpdateFloat(value)
+	})
+}
+
+func (rs *RunSummary) updateSummaryInt(
+	path pathtree.TreePath,
+	value int64,
+) (*service.SummaryItem, error) {
+	return rs.updateSummary(path, func(ms *metricSummary) {
+		ms.UpdateInt(value)
+	})
+}
+
+func (rs *RunSummary) updateSummaryOther(
+	path pathtree.TreePath,
+	value any,
+) (*service.SummaryItem, error) {
+	return rs.updateSummary(path, func(ms *metricSummary) {
+		ms.UpdateOther(value)
+	})
+}
```

Comment:
do we really need these helper functions if it is just a one liner?

Reply:
nope, you're right! removed them